### PR TITLE
 Support reopening single files via unix domain socket in DaemonApp

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
`DaemonApp`'s `ReopenableFilesExt` now supports reopening commands from the
UnixSocketExt. If you have the UnixSocketExt bound to the socket, sending command
`reopen_logfile file_path` will reopen this file, if registered, and acknowledge
the action with `ACK` or `ERROR` strings.

This is different than reopening files with the reopen signal, which
is not capable passing the file path to it, so it forces reopening all
files.

### CODE REVIEW

**Comment by nemanja-boric-sociomantic** [comment]
```
echo reopen_logfile /srv/dlsnode/log/root.log | sudo nc -U /srv/dlsnode/dlsnode.socket
ACK
```

File: relnotes/reopensocket.feature.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+* `ocean.util.app.DaemonApp`
+
+  `DaemonApp`'s `ReopenableFilesExt` now supports reopening commands from the
+  UnixSocketExt. If you have the UnixSocketExt bound to the socket, sending command
+  `reopen_logfile file_path` will reopen this file, if registered, and acknowledge
```
Comment (gavin-norman-sociomantic) [inline relnotes/reopensocket.feature.md]: "will reopen the specified file"

**Comment by nemanja-boric-sociomantic** [comment]
With this in place, reopening only affected files with the logrotate can be done as follows:

1. use `nosharedscripts` directive, so the post-rotate script is called once per file, with the absolute file path passed to it.
2. From the absolute file path, get the log directory and deduce socket placement from it (we could have convention that the socket is in the application's `run/` directory, so the socket path would be `
SOCKETPATH=$(dirname $LOGPATH)/../run/application.socket`
3. Simply execute:

```
test $(echo reopen_logfile $LOGPATH | sudo nc -U $SOCKETPATH) = ACK
```

and that should rotate your single logfile.

**Comment by gavin-norman-sociomantic** [comment]
https://github.com/sociomantic-tsunami/ocean/pull/314#issuecomment-337876590 would be super useful in the release notes. We should try to make this feature as easy as possible for people to start using.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the extended release notes.

File: src/ocean/util/app/ext/ReopenableFilesExt.d [hunk 2]
```diff
@@ -70,14 +93,27 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
 
     ***************************************************************************/
 
-    public this ( SignalExt signal_ext = null, int reopen_signal = SIGHUP )
+    public this ( SignalExt signal_ext = null,
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/app/ext/ReopenableFilesExt.d]: Oh wait... this is a breaking change, or?

**Review by gavin-norman-sociomantic — COMMENTED** [review]
Submitting an early review because of the breaking change question.

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 2 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Puh, yeah, I've missed that - I originally targeted v4.x.x. Just having new constructor with suffice, though. Will update.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the dodged breaking change.

File: src/ocean/io/device/File.d [hunk 3]
```diff
@@ -412,6 +412,18 @@ class File : Device, Device.Seek, Device.Truncate
...
+
+            Returns:
+                path used by this file.
+
+        **********************************************************************/
+
+        public cstring path ()
+        {
+            return this.path_;
+        }
```
Comment (leandro-lucarella-sociomantic) [inline src/ocean/io/device/File.d:425]: Why isn't this done in `toString()` too?

File: relnotes/reopensocket.feature.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
+* `ocean.util.app.DaemonApp`
+
+  `DaemonApp`'s `ReopenableFilesExt` now supports reopening commands from the
+  UnixSocketExt. If you have the UnixSocketExt bound to the socket, sending
+  command `reopen_logfile file_path` will reopen the specified file, if
+  registered, and acknowledge the action with `ACK` or `ERROR` strings.
```
Comment (leandro-lucarella-sociomantic) [inline relnotes/reopensocket.feature.md]: I wonder if, for debugging purposes, a message should be returned with the `ERROR` too (like `ERROR <free text>`).

File: relnotes/reopensocket.feature.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  with logrotate can be done as follows:
+
+    1) use `nosharedscripts` directive, so the post-rotate script is called
+       once per file, with the absolute file path passed to it.
+    2) From the absolute file path, get the log directory and deduce socket
+       location from it (application could place the socket in its `run/`
+       directory, so the socket path would be
+       `SOCKETPATH=$(dirname $LOGPATH)/../application.socket`.
+    3) Finally, in post-rotate script, simply executing
+       `test $(echo reopen_logfile $LOGPATH | nc -q 2 -U $SOCKETPATH) = ACK` will
```
Comment (leandro-lucarella-sociomantic) [inline relnotes/reopensocket.feature.md]: You probably should add quotes to `$(echo ...)`, otherwise it could be an empty string and make `test` give an error, right?

File: src/ocean/util/app/DaemonApp.d [hunk 6]
```diff
@@ -380,6 +375,12 @@ public abstract class DaemonApp : Application,
         this.config_ext.registerExtension(this.unix_socket_ext);
         this.registerExtension(this.unix_socket_ext);
 
+        // Create and register repoenable files extension
+        this.reopenable_files_ext = new ReopenableFilesExt(this.signal_ext,
+            this.unix_socket_ext,
+            settings.reopen_signal);
+        this.registerExtension(this.reopenable_files_ext);
+
```
Comment (leandro-lucarella-sociomantic) [inline src/ocean/util/app/DaemonApp.d]: Why is this now registered twice?

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 2 above)
Comment (leandro-lucarella-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Ah, I guess it's registered twice because of this... Maybe add an issue and a comment to clean this up in a following major?

File: src/ocean/util/app/ext/ReopenableFilesExt.d [hunk 7]
```diff
@@ -70,14 +94,54 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
...
+            reopen_command = if unix socket ext is non-null, command to trigger
+                reopenning of the registered file (passed as the first
+                argument to the command).
+
+    ***************************************************************************/
+
+    public this ( SignalExt signal_ext,
+        UnixSocketExt unix_socket_ext,
+        int reopen_signal = SIGHUP,
+        cstring reopen_command = "reopen_logfile")
```
Comment (leandro-lucarella-sociomantic) [inline src/ocean/util/app/ext/ReopenableFilesExt.d]: Do we really want to make this configurable? It seems strange to be able to introduce inconsistencies in the socket interface...

File: src/ocean/util/app/ext/ReopenableFilesExt.d [hunk 8]
```diff
@@ -130,6 +224,27 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
...
+            args = list of arguments received from the socket - should contain
+                the filename of the log to rotate.
+            send_response = delegate to send the response to the client
+
+    *****************************************************************************/
+
+    private void socketReloadCommand ( cstring[] args,
+            void delegate ( cstring response ) send_response )
+    {
+        // Todo: reopen the file here.
```
Comment (leandro-lucarella-sociomantic) [inline src/ocean/util/app/ext/ReopenableFilesExt.d]: I guess this `TODO` is outdated?

File: src/ocean/io/device/File.d (hunk 3 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/io/device/File.d:425]: Because `toString()` must return `istring` (from `Object`).

File: src/ocean/util/app/DaemonApp.d (hunk 6 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/DaemonApp.d]: Probably accident - we need to register it only once. Will fix.

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 7 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Well, we do the same thing for the SIGHUP. Note that people don't actually use this class directly, but it's used withing DaemonApp.

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 8 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Oh, right.

File: relnotes/reopensocket.feature.md (hunk 5 above)
Comment (nemanja-boric-sociomantic) [reply relnotes/reopensocket.feature.md]: Yeah.

Comment (nemanja-boric-sociomantic) [reply relnotes/reopensocket.feature.md]: Good catch!

File: relnotes/reopensocket.feature.md (hunk 4 above)
Comment (nemanja-boric-sociomantic) [reply relnotes/reopensocket.feature.md]: I could return ERROR + the error text, no problem.

File: src/ocean/util/app/DaemonApp.d (hunk 6 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/DaemonApp.d]: Wait, where do you see it's registered twice?

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/DaemonApp.d]: I've just moved it to be registered after UnixSocketExt, since we need it for registering the handler.

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

File: src/ocean/util/app/DaemonApp.d (hunk 6 above)
Comment (leandro-lucarella-sociomantic) [reply src/ocean/util/app/DaemonApp.d]: I guess I was colorblind and didn't see the red above... All good.

**Review by leandro-lucarella-sociomantic — APPROVED** [review]
All the pending stuff is minor, so approving, feel free to merge and then work on any improvements.

**Comment by mathias-lang-sociomantic** [comment]
Any chance we could get this tested by the right people ( :wave: @andrew-stevenson-sociomantic ) so we are sure it fulfills their use-case ? :)
Thanks a lot for taking care of this @nemanja-boric-sociomantic 
Fixes https://github.com/sociomantic-tsunami/ocean/issues/190

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 7 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Travis says this should be an `istring`
Also this works with *any* file opened with `ReusableFilesExt` right ? In this case `logfile` might be a bit too restrictive of a name.

File: src/ocean/util/app/ext/ReopenableFilesExt.d [hunk 9]
```diff
@@ -130,6 +224,37 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
...
+    private void socketReloadCommand ( cstring[] args,
+            void delegate ( cstring response ) send_response )
+    {
+        if (args.length == 0)
+        {
+            send_response("ERROR: missing name of the file to rotate.\n");
+            return;
+        }
+
+        auto filename = args[0];
```
Comment (mathias-lang-sociomantic) [inline src/ocean/util/app/ext/ReopenableFilesExt.d]: Could we use a `while` loop here, so one can pass multiple files ?

File: src/ocean/util/app/ext/ReopenableFilesExt.d [hunk 10]
```diff
@@ -70,14 +94,54 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
...
+
+    public this ( SignalExt signal_ext,
+        UnixSocketExt unix_socket_ext,
+        int reopen_signal = SIGHUP,
+        cstring reopen_command = "reopen_logfile")
     {
+        this.cwd = Environment.directory();
+
+        // Environment.directory makes sure for the trailing /, confirm it
+        verify (this.cwd[$-1] == '/');
```
Comment (mathias-lang-sociomantic) [inline src/ocean/util/app/ext/ReopenableFilesExt.d]: Extra space

Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Also the comment sounds weird to me

File: relnotes/reopensocket.feature.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    1) use `nosharedscripts` directive, so the post-rotate script is called
+       once per file, with the absolute file path passed to it.
+    2) From the absolute file path, get the log directory and deduce socket
+       location from it (application could place the socket in its `run/`
+       directory, so the socket path would be
+       `SOCKETPATH=$(dirname $LOGPATH)/../application.socket`.
+    3) Finally, in post-rotate script, simply executing
+       `test "$(echo reopen_logfile $LOGPATH | nc -U $SOCKETPATH)" = ACK` will
+       instruct the application to reopen the given logfile and report status
+       to logrotate.
```
Comment (mathias-lang-sociomantic) [inline relnotes/reopensocket.feature.md]: Maybe give a very simple sample file ?

File: src/ocean/util/app/ext/ReopenableFilesExt.d [hunk 12]
```diff
@@ -70,14 +94,54 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
 
     ***************************************************************************/
 
-    public this ( SignalExt signal_ext = null, int reopen_signal = SIGHUP )
+    public this ( SignalExt signal_ext = null,
+        int reopen_signal = SIGHUP )
```
Comment (mathias-lang-sociomantic) [inline src/ocean/util/app/ext/ReopenableFilesExt.d]: Why the change indent ?

Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Since this is internal-only, should it be deprecated ?

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 7 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Another comment about this: It looks weird to have signal-related thing, unixsocket, signal-related, unix-socket.

I understand why you went with this, and the alternative is not too appealing as well, so I'm going to propose an alternate solution: since both `SignalExt` and `UnixSocketExt` are optional and independent of each other, why don't we move the logic to `setup` methods, e.g. `setupUnixSocket(UnixSocketExt, istring = "reopen_file")` and `setupSignalExt(SignalExt, int = SIGHUP)` (naming up for discussion).

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 9 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: I didn't anticipated that one can do that, since logrotate doesn't do it, but it makes sense, I'll update it.

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 12 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Interesting thing that this is not internal only - once could create their completely separate application hierarchy and use this. However, I will deprecate this, good catch :+1:

Indentation was changed via copy/pasting from other constructor. Will fix.

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 10 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Yes, should be `Environment.directory should always return trailing /, confirm that that's the case`

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 7 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Right, deprecating constructor and having setup methods replace it would work, good idea :+1:

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, adding some test would be nice. It is a bit annoying but it should be possible doing something like: create extension, register file X, write, move file externally to X and create a new file X, write again, check it's written to Y (and not to X), reopen files, write again and check it's written to X (and not to Y).

**Comment by nemanja-boric-sociomantic** [comment]
Yeah, the test could/should be added.

File: src/ocean/util/app/ext/ReopenableFilesExt.d [hunk 13]
```diff
@@ -114,6 +178,36 @@ public class ReopenableFilesExt : IApplicationExtension, ISignalExtExtension
...
+        Returns:
+            true if the file was registered and reopen, false otherwise.
+
+    ***************************************************************************/
+
+    public bool reopenFile ( cstring file_path )
+    {
+        foreach (file; this.open_files)
+        {
+            path_buffer.length = 0;
```
Comment (mihails-strasuns-sociomantic) [inline src/ocean/util/app/ext/ReopenableFilesExt.d]: I'd recommend `path_buffer.reset()`, it makes it more obvious you don't need `enableStomping` :)

Comment (mathias-lang-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Yeah I started typing `enableStomping` and had the "oh..." moment :P

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: The sole purpose of Buffer :-P

File: src/ocean/util/app/ext/ReopenableFilesExt.d (hunk 7 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/app/ext/ReopenableFilesExt.d]: Damn, I can't deprecate the existing constructor as all the methods are optional, so simply `auto ext = new ReopenableFilesExt` will trigger the deprecation :(.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with all points except the example logrotate file and test.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the test.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the example logrotate config file and the fact that there may be multiple files registered with the same filename, and they all needs to be reloaded.

**Comment by nemanja-boric-sociomantic** [comment]
I think this includes all requested changes on this PR.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with the last bit: `ReopenableFilesExt` was getting the file's path via `File.toString` which would always allocate, given it needs to return `istring`. It's now tweaked to use newly added `File.path`.

**Review by mihails-strasuns-sociomantic — APPROVED** [review]

**Review by mathias-lang-sociomantic — APPROVED** [review]
LGTM, pending testing I assume ?

**Comment by nemanja-boric-sociomantic** [comment]
Yes, I've tested it with the DLS node and it works. I wonder if @gautam-kotian-sociomantic could help us test this as well, if @andrew-stevenson-sociomantic doesn't have much time.

If nobody has time for a test, I think it's fair to merge this as is, given that I did a bit of testing, and fix issues in the patch releases, than to leave this PR rotting.

**Comment by andrew-stevenson-sociomantic** [comment]
> if @andrew-stevenson-sociomantic doesn't have much time.

I suspect it will take a lot of time for me to test as I would need to update ocean, add command socket support to an app, switch from our logging stuff to ocean and then test :-(

**Comment by nemanja-boric-sociomantic** [comment]
Just a note there: adding command socket support is done automatically with the recent oceans (just add `[UNIX_SOCKET]path=app.path` in the config file). But yeah, removing the custom stuff is more problematic.

**Comment by leandro-lucarella-sociomantic** [comment]
If we have unit/integration tests and simple test seems to work, I would just merge.

**Comment by nemanja-boric-sociomantic** [comment]
Yes, the intergration test has been added.

**Comment by leandro-lucarella-sociomantic** [comment]
I don't get why the coverage is so crappy now. Lately I couldn't make sense of the coverage reports :angry: 

Also, needs rebase now.

**Comment by nemanja-boric-sociomantic** [comment]
The biggest problem we have IMHO is that the coverage is reported file per file only. So, having something tested in the different file (like we're doing in the integration tests), doesn't help coverage at all :(. The second, separate problem, which is way more annoying is that these unexpected coverage changes. I wonder if that's related to the fact that we started uploading coverage for all matrix items (and actually building stuff with F=production), so codecov is going crazy because of that.

**Comment by nemanja-boric-sociomantic** [comment]
Rebased.

**Comment by leandro-lucarella-sociomantic** [comment]
> I wonder if that's related to the fact that we started uploading coverage for all matrix items (and actually building stuff with F=production), so codecov is going crazy because of that.

Mmm, maybe it is worth trying to disable codecov for F=production builds then. It's really annoying right now.
